mods/per-app-language-preferences.wh.cpp

```diff
@@ -65,30 +65,42 @@ Note that it changes the language only and does not affect the encoding used in
 
 // ===========================================================
 
+bool langid_determined = false;
+int my_langid = 0;
+
 using GetUserDefaultUILanguage_t = short(WINAPI*)();
 GetUserDefaultUILanguage_t GetUserDefaultUILanguage_Original;
 short WINAPI GetUserDefaultUILanguage_Hook() {
-    wchar_t filename_buf[2049];
-
-    Wh_Log(L">GetUserDefaultUILanguage");
-
-    GetModuleFileNameW(NULL, filename_buf, 2049);
-    Wh_Log(L">Process file: %ls", filename_buf);
-
-    for(int index = 0; ; index++) {
-        PCWSTR glob = Wh_GetStringSetting(L"programList[%d].glob", index);
-        if(!*glob) {
+    // Wh_Log(L">GetUserDefaultUILanguage");
+
+    if(!langid_determined) {
+        wchar_t *filename_buf = new wchar_t[2048];
+        GetModuleFileNameW(NULL, filename_buf, 2048);
+        Wh_Log(L">Process file: %ls", filename_buf);
+
+        int lang_id = -1;
+        for(int index = 0; ; index++) {
+            PCWSTR glob = Wh_GetStringSetting(L"programList[%d].glob", index);
+            if(!*glob) {
+                Wh_FreeStringSetting(glob);
+                break;
+            }
+            if(S_OK == PathMatchSpecExW(filename_buf, glob, PMSF_MULTIPLE)) {
+                lang_id = Wh_GetIntSetting(L"programList[%d].langId", index);
```

Comment:
The string is not freed with `Wh_FreeStringSetting` in this case. Note that you can also use `WindhawkUtils::StringSetting` to have it freed automatically.